src/services/Metadata.ts

```diff
@@ -0,0 +1,295 @@
+import type { X509Certificate } from '@peculiar/x509'
+
+import { AlgorithmIdentifier, Certificate, ContentInfo, EncapsulatedContentInfo, IssuerAndSerialNumber, SignedData, SignerInfo } from 'pkijs'
+import { base64ToBuffer, bufferToBase64, bufferToString, stringToBuffer } from './bufferUtils.ts'
+import { compress, uncompress } from './compression.ts'
+import { encryptWithAES, sha256Hash } from './crypto.ts'
+import logger from './logger.ts'
+import { convertEncryptionKeyToSigningKey } from './privateKeyUtils.ts'
+
+interface IRawMetadataUser {
+	/**
+	 * The user's ID
+	 */
+	userId: string
+	/**
+	 * The user's public key (x509 as PEM)
+	 */
+	certificate: string
+	/**
+	 * The metadata key encrypted with the user's public key, then base64 encoded.
+	 * So the user can decrypt it with their private key and use it to decrypt the metadata.
+	 */
+	encryptedMetadataKey: string
+}
+
+interface IRawMetadataFiledrop {
+	ciphertext: string
+	nonce: string
+	authenticationTag: string
+	users: Array<{
+		userId: string
+		encryptedFiledropKey: string
+	}>
+}
+
+interface IRawMetadata {
+	metadata: {
+		/**
+		 * Encrypted metadata (AES/GCM/NoPadding, 128 bit key size with key from "users" section).
+		 *
+		 * Data is first gzipped, then encrypted, then base64 encoded.
+		 */
+		ciphertext: string
+		nonce: string // sometimes also called iv
+		authenticationTag: string
+	}
+
+	/**
+	 * Array of users who have access to the folder
+	 */
+	users: IRawMetadataUser[]
+
+	/**
+	 * Mapping of uuid to filedrop info
+	 */
+	filedrop: Record<string, IRawMetadataFiledrop>
+
+	/**
+	 * Metadata version
+	 */
+	version: string
+}
+
+interface IMetadataFile {
+	/** Unencrypted file name */
+	filename: string
+	/** Mimetype. If unknown, use "application/octet-stream" */
+	mimetype: string
+	/** Encryption algorithm: AES/GCM/NoPadding (128 bit key size) */
+	nonce: string
+	authenticationTag: string
+	key: string
+}
+
+interface IMetadata {
+	/** list of hashes of metadata-keys */
+	keyChecksums: string[]
+	deleted: boolean
+	counter: number
+	/**
+	 * Mapping of uuids to cleartext folder names
+	 */
+	folders: Record<string, string>
+	files: Record<string, IMetadataFile>
+}
+
+export class Metadata {
+	#metadata?: IMetadata
+	#metadataKey: CryptoKey
+	#users: IRawMetadataUser[]
+	#filedrop: Record<string, IRawMetadataFiledrop>
+	#version = '2.0' // TODO: Should we support multiple versions?
```

Comment:
I assume that most E2EE folders are migrated to v2 now. So we can make our life easier and only support `2.0`. Maybe display a warning if we encounter a non `2.0` metadata.

Reply:
Yes it was meant as a reminder for me to gracefully handle older (or newer?) versions